### PR TITLE
Feature/nti notice to improve  create

### PR DESCRIPTION
**Pull request by Menol-LK**
**What is the change?**
Notice to improve - create new action

**Why do we need the change?**
New feature

**What is the impact?**
Most of the changes are new, with some changes to existing code

**Azure DevOps Ticket**
103232

### CODE REVIEW

File: ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/AddConditions.cshtml.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+		public long? NtiId { get; set; }
+
+		public ICollection<NtiConditionModel> SelectedConditions { get; private set; }
+		public ICollection<NtiConditionDto> AllConditions { get; private set; }
+
+		public AddConditionsPageModel(INtiWarningLetterStatusesCachedService ntiWarningLetterStatusesCachedService,
+			INtiWarningLetterReasonsCachedService ntiWarningLetterReasonsCachedService,
+			INtiModelService ntiModelService,
+			INtiConditionsCachedService ntiConditionsCachedService,
+			ILogger<NtiWarningLetterConditionModel> logger)
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/AddConditions.cshtml.cs:47]: Wrong class in logger, is NtiWarningLetterConditionModel, should be AddConditionsPageModel

File: ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/AddConditions.cshtml.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
+﻿using ConcernsCaseWork.Mappers;
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/AddConditions.cshtml.cs:1]: Some unused usings to tidy up

File: ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/AddConditions.cshtml.cs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+		private void ExtractCaseUrnFromRoute()
+		{
+			if (TryGetRouteValueInt64("urn", out var caseUrn))
+			{
+				CaseUrn = caseUrn;
+			}
+			else
+			{
+				throw new InvalidOperationException("CaseUrn not found in the route");
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/AddConditions.cshtml.cs:140]: I've seen both InvalidOperationException and Exception thrown when failing to get CaseUrn from route. Is InvalidOperationException the preferred going forward?

File: ConcernsCaseWork/ConcernsCaseWork/Services/Nti/INtiModelService.cs [hunk 4]
```diff
@@ -6,6 +6,13 @@ namespace ConcernsCaseWork.Services.Nti
 {
 	public interface INtiModelService
 	{
-		public Task<ICollection<NtiModel>> GetNtisForCase(long caseUrn);
+		Task<ICollection<NtiModel>> GetNtisForCase(long caseUrn);
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork/Services/Nti/INtiModelService.cs:9]: I've seen some methods that return a task suffixed with 'Async'. Is there a preferred standard for this? I've been adding a suffix if I've seen one is missing.

File: ConcernsCaseWork/ConcernsCaseWork/Services/Nti/NtiModelService.cs [hunk 5]
```diff
@@ -16,10 +17,52 @@ public NtiModelService(INtiCachedService ntiCachedService)
...
 
+		public async Task<NtiModel> CreateNtiAsync(NtiModel ntiModel)
+		{
+			var created = await _ntiCachedService.CreateNtiAsync(NtiMappers.ToDBModel(ntiModel));
+			return NtiMappers.ToServiceModel(created);
+		}
+
+		public async Task<NtiModel> GetNtiAsync(string continuationId)
+		{
+			if (string.IsNullOrWhiteSpace(continuationId))
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork/Services/Nti/NtiModelService.cs:28]: Are we considering guard clauses going forward, or is this still being decided?

File: ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/Case/Management/Action/Nti/AddPageModelTests.cs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+			var pageModel = SetupAddPageModel(mockNtiModelService, mockNtiReasonsService, mockNtiStatusesService, mockLogger);
+
+			var routeData = pageModel.RouteData.Values;
+			routeData.Add("urn", 1);
+
+			// act
+			await pageModel.OnGetAsync();
+
+			// assert
+			mockLogger.Verify(
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/Case/Management/Action/Nti/AddPageModelTests.cs:61]: This test doesn't actually test what it says it tests - it says it is testing that a PageModel is returned but actually the page could throw an exception and this test would still pass.
It is only testing that a message is logged.

File: ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/Case/Management/Action/Nti/AddPageModelTests.cs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Primitives;
+using Moq;
+using NUnit.Framework;
+using Service.Redis.Nti;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace ConcernsCaseWork.Tests.Pages.Case.Management.Action.Nti
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork.Tests/Pages/Case/Management/Action/Nti/AddPageModelTests.cs:19]: There's quite a lot of logic in the Add page - could do with some more tests maybe?

File: ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/AddConditions.cshtml.cs [hunk 8]
```diff
@@ -0,0 +1,173 @@
+﻿using ConcernsCaseWork.Mappers;
+using ConcernsCaseWork.Models;
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/AddConditions.cshtml.cs:2]: There aren't any tests for AddConditions.cshtml.cs?

File: ConcernsCaseWork/Service.Redis/Nti/INtiCachedService.cs [hunk 9]
```diff
@@ -8,6 +8,11 @@ namespace Service.Redis.Nti
 {
 	public interface INtiCachedService
```
Comment (emma-w-nimble) [inline ConcernsCaseWork/Service.Redis/Nti/INtiCachedService.cs:9]: Are we no longer doing service tests? (Service.Redis.Tests)

File: ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/AddConditions.cshtml.cs (hunk 3 above)
Comment (Menol-LK) [reply ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/AddConditions.cshtml.cs:140]: Yes, InvalidOperationException is the most suitable exception to be thrown here.

File: ConcernsCaseWork/ConcernsCaseWork/Services/Nti/NtiModelService.cs (hunk 5 above)
Comment (Menol-LK) [reply ConcernsCaseWork/ConcernsCaseWork/Services/Nti/NtiModelService.cs:28]: This story was already being implemented when the discussion started. Will start using guard clauses going forward.

**Review by chrisdexnimble — APPROVED** [review]
Approved with comments, review when you get time.
Somethings might make a good topic to talk about in our weekly tech meeting.

File: ConcernsCaseWork/ConcernsCaseWork/Mappers/NtiMappers.cs [hunk 10]
```diff
@@ -1,16 +1,80 @@
 ﻿using ConcernsCaseWork.Models.CaseActions;
 using Service.TRAMS.Nti;
+using System;
+using System.Linq;
 
 namespace ConcernsCaseWork.Mappers
 {
 	public static class NtiMappers
 	{
```
[the remark this answers is not in the document]
Comment (Menol-LK) [reply ConcernsCaseWork/ConcernsCaseWork/Mappers/NtiMappers.cs:9]: yes, this story started development before we discussed automapper. will use that going forward.

File: ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/Add.cshtml.cs [hunk 11]
```diff
@@ -11,153 +12,296 @@
...
+			ntiModel.CaseUrn = newValues.CaseUrn;
+			ntiModel.Reasons = newValues.Reasons;
+			ntiModel.Status = newValues.Status;
+			ntiModel.Notes = newValues.Notes;
+			ntiModel.SentDate = newValues.SentDate;
+			ntiModel.CreatedAt = newValues.CreatedAt;
+			ntiModel.UpdatedAt = newValues.UpdatedAt;
+
+			return ntiModel;
 		}
```
[the remark this answers is not in the document]
Comment (Menol-LK) [reply ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/Add.cshtml.cs:261]: yes, one more thing we agreed after this story started. model binding will be used in new stories.

**Pull request merged**